tracer/src/Datadog.Trace/Ci/Ipc/IpcDualChannel.cs

```diff
@@ -0,0 +1,82 @@
+// <copyright file="IpcDualChannel.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+#nullable enable
+
+using System;
+using Datadog.Trace.Vendors.Newtonsoft.Json;
+using Datadog.Trace.Vendors.Newtonsoft.Json.Serialization;
+
+namespace Datadog.Trace.Ci.Ipc;
+
+internal abstract class IpcDualChannel : IDisposable
+{
+    private readonly IChannel _recvChannel;
+    private readonly IChannelReceiver _recvChannelReceiver;
+    private readonly IChannel _sendChannel;
+    private readonly IChannelWriter _sendChannelWriter;
+    private readonly JsonSerializerSettings _serializerSettings;
+
+    protected IpcDualChannel(string recvName, string sendName)
+    {
+        _recvChannel = new CircularChannel(recvName);
+        _recvChannelReceiver = _recvChannel.GetReceiver();
+        _recvChannelReceiver.MessageReceived += OnMessageReceived;
+
+        _sendChannel = new CircularChannel(sendName);
+        _sendChannelWriter = _sendChannel.GetWriter();
+
+        _serializerSettings = new JsonSerializerSettings()
+        {
+            TypeNameHandling = TypeNameHandling.All,
+            Formatting = Formatting.None,
+            NullValueHandling = NullValueHandling.Ignore,
+            SerializationBinder = new CustomSerializationBinder(),
+        };
+    }
+
+    public event EventHandler<object>? MessageReceived;
+
+    private void OnMessageReceived(object? sender, byte[] data)
+    {
+        var jsonMessage = Util.EncodingHelpers.Utf8NoBom.GetString(data);
+        var message = JsonConvert.DeserializeObject(jsonMessage, _serializerSettings);
+        if (message != null)
+        {
+            MessageReceived?.Invoke(sender, message);
+        }
+    }
+
+    public bool TrySendMessage(object message)
+    {
+        var jsonMessage = JsonConvert.SerializeObject(message, _serializerSettings);
+        var bytes = Util.EncodingHelpers.Utf8NoBom.GetBytes(jsonMessage);
+        return _sendChannelWriter.TryWrite(bytes);
+    }
+
+    public void Dispose()
+    {
+        _recvChannel.Dispose();
+        _sendChannel.Dispose();
+    }
+
+    private class CustomSerializationBinder : ISerializationBinder
+    {
+        public Type BindToType(string? assemblyName, string typeName)
```

Comment:
yea, primitive types are already covered, this is a protection to avoid creating any object instance from a `$type` variable coming our of proc. A well known attack vector for serializers, in this case the protection is kind on naive I know, only be able to create objects from the Datadog.Trace.* assemblies.

Reply:
Yup, totally get the attack PoV, no arguments there 👍 

What about types that aren't primitives, but are in the BCL (for example). If this restriction works for now then it's fine by me, just want to make sure we're not missing anything 🙂